src/test/tuning/tuner_test.cpp

```diff
@@ -0,0 +1,164 @@
+  tuner_eval.set_selector(std::make_unique<MockSelector>(1, RuntimeMsList{1, 1, 1}));
+
+  tuner_eval.set_time_budget(0.1);  // in seconds
+
+  tuner_eval.schedule_tuning_process();
+  tuner_eval.wait_for_completion();
+
+  EXPECT_EQ(tuner_eval.status(), Tuner::Status::Timeout);
+
+  // during selection phase
+  Tuner tuner_sel;
+  tuner_sel.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner_sel.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner_sel.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner_sel.set_selector(std::make_unique<MockSelector>(200, RuntimeMsList{1, 1, 1}));
+
+  tuner_sel.set_time_budget(0.1);  // in seconds
+
+  tuner_sel.schedule_tuning_process();
+  tuner_sel.wait_for_completion();
+
+  EXPECT_EQ(tuner_sel.status(), Tuner::Status::Timeout);
+
+  // during execution phase
+  Tuner tuner_exec;
+  tuner_exec.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner_exec.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner_exec.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner_exec.set_selector(std::make_unique<MockSelector>(1, RuntimeMsList{1, 200, 1}));
+
+  tuner_exec.set_time_budget(0.1);  // in seconds
+
+  tuner_exec.schedule_tuning_process();
+  tuner_exec.wait_for_completion();
+
+  EXPECT_EQ(tuner_exec.status(), Tuner::Status::Timeout);
+}
+
+TEST_F(TunerTest, TuningEvaluationTimeout) {
+  Tuner tuner;
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(200));
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner.set_selector(std::make_unique<MockSelector>(200, RuntimeMsList{1, 1, 1}));
+
+  tuner.set_time_budget(0.3, Tuner::NoBudget, 0.1, Tuner::NoBudget);  // in seconds
+
+  tuner.schedule_tuning_process();
+  tuner.wait_for_completion();
+
+  EXPECT_EQ(tuner.status(), Tuner::Status::EvaluationTimeout);
+}
+
+TEST_F(TunerTest, TuningSelectionTimeout) {
+  Tuner tuner;
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner.set_selector(std::make_unique<MockSelector>(200, RuntimeMsList{1, 200, 1}));
+
+  tuner.set_time_budget(0.3, Tuner::NoBudget, Tuner::NoBudget, 0.1);  // in seconds
+
+  tuner.schedule_tuning_process();
+  tuner.wait_for_completion();
+
+  EXPECT_EQ(tuner.status(), Tuner::Status::SelectionTimeout);
+}
+
+TEST_F(TunerTest, TuningExecutionTimeout) {
+  Tuner tuner;
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner.add_evaluator(std::make_unique<MockEvaluator>(1));
+  tuner.set_selector(std::make_unique<MockSelector>(1, RuntimeMsList{200, 200, 200}));
+
+  tuner.set_time_budget(0.3, 0.1);  // in seconds
+
+  tuner.schedule_tuning_process();
+  tuner.wait_for_completion();
+
+  EXPECT_EQ(tuner.status(), Tuner::Status::ExecutionTimeout);
+}
+
+}  // namespace opossum
```

Comment:
I like how you evaluate the different components separately. Additionally, I think we should have an end-to-end test that makes sure that the index selector actually does something

Reply:
Definitely